src/coreclr/tools/Common/TypeSystem/IL/Stubs/PInvokeILEmitter.cs

```diff
@@ -231,16 +233,23 @@ private void EmitDelegateCall(DelegateMarshallingMethodThunk delegateMethod, PIn
 
         private void EmitPInvokeCall(PInvokeILCodeStreams ilCodeStreams)
         {
-            if (!_flags.PreserveSig && _targetMethod.Signature.ReturnType != _targetMethod.Context.GetWellKnownType(WellKnownType.Void))
-                throw new NotSupportedException();
+            if (!_flags.PreserveSig)
+            {
+                TypeDesc managedReturnType = _targetMethod.Signature.ReturnType;
+                bool supportedType = managedReturnType.IsPrimitive || managedReturnType.IsInterface || managedReturnType.IsObject;
+                if (!supportedType)
```

Comment:
What prevents the other cases from being supported? I would expect the implementation to work for all marshalers without special casing.

Reply:
It require more testing. Nothing in principle, just my inability to see all supported cases.
Also https://github.com/dotnet/runtime/blob/1843cdef72e1e7c810c4ea85c677d679ff3b66f4/src/coreclr/vm/dllimport.cpp#L3560-L3564 indicate cases which are not supported, but that does not give me much information what is supported. I think I can add Enum and String without probably much problems if needed.

Reply:
All types that can be marshalled should be supported. 

The unsupported cases in CoreCLR are just a legacy baggage from the times where the marshalling stubs were hand-emitted x86 assembly and handling them would require writing a pile of new hand-emitted x86 assembly. We should not need need to worry about artificially throwing exceptions for the cases that are unsupported in CoreCLR for no good reason. 

I think that the transformation for PreserveSig = false should be very regular, it should depend on the regular `[out]` marshalling. It should not depend on the actual type at all.


```
[PreserveSig = false]
RETVAL foo(...);
```
should be transformed to:
```
RETVAL foo(...)
{
   RETVAL retNative = default;

   marshal everything usual

   actual_call(..., &retNative);

   unmarshal everything as usual (including retNative / retManaged)

   return retManaged;
}
```